site/lib/src/components/pages/changelog/changelog_filters_sidebar.dart

```diff
@@ -0,0 +1,331 @@
+      selectedAreas.add(area);
+    } else {
+      selectedAreas.remove(area);
+    }
+    notifyListeners();
+  }
+
+  void setVersion(String version, bool isSelected) {
+    if (!version.contains('.')) {
+      // It's a major version (e.g. "3")
+      if (isSelected) {
+        selectedVersions.add(version);
+        // Select all versions starting with this major
+        final majorPrefix = '$version.';
+        selectedVersions.addAll(
+          availableVersions.where((v) => v.startsWith(majorPrefix)),
+        );
+      } else {
+        selectedVersions.remove(version);
+        // Deselect all versions starting with this major
+        final majorPrefix = '$version.';
+        selectedVersions.removeAll(
+          availableVersions.where((v) => v.startsWith(majorPrefix)),
+        );
+      }
+    } else {
+      if (isSelected) {
+        selectedVersions.add(version);
+        // Check if all versions of this major are selected
+        final major = version.split('.').first;
+        final majorPrefix = '$major.';
+        final allMajor = availableVersions.where(
+          (v) => v.startsWith(majorPrefix),
+        );
+        if (allMajor.every(selectedVersions.contains)) {
+          selectedVersions.add(major);
+        }
+      } else {
+        selectedVersions.remove(version);
+        // Deselect the major checkbox if it was selected
+        final major = version.split('.').first;
+        selectedVersions.remove(major);
+      }
+    }
+    notifyListeners();
+  }
+
+  void reset() {
+    selectedTags.clear();
+    selectedAreas.clear();
+    selectedVersions.clear();
+    notifyListeners();
+  }
+
+  void disposeState({bool notify = true}) {
+    selectedTags.clear();
+    selectedAreas.clear();
+    selectedVersions.clear();
+    availableAreas.clear();
+    availableVersions.clear();
+    if (notify) {
+      notifyListeners();
+    }
+  }
+
+  int compareVersions(String v1, String v2) {
+    final p1 = v1.split('.').map(int.parse).toList();
+    final p2 = v2.split('.').map(int.parse).toList();
+
+    for (var i = 0; i < max(p1.length, p2.length); i++) {
+      final n1 = i < p1.length ? p1[i] : 0;
+      final n2 = i < p2.length ? p2[i] : 0;
+      if (n1 != n2) return n1.compareTo(n2);
+    }
+    return 0;
+  }
+
+  int compareAreas(String a, String b) {
+    if (a == b) return 0;
+    // 1. SDK at the top
+    if (a == 'SDK') return -1;
+    if (b == 'SDK') return 1;
+    // 2. Language second
+    if (a == 'Language') return -1;
+    if (b == 'Language') return 1;
+    // 3. Docs at the bottom
+    if (a == 'Docs') return 1;
+    if (b == 'Docs') return -1;
+    // 4. Everything else alphabetical
+    return a.compareTo(b);
+  }
+
+  List<ChangelogEntry> filterEntries(
+    List<ChangelogEntry> entries,
+    String searchQuery,
+  ) {
+    if (searchQuery.isEmpty &&
+        selectedTags.isEmpty &&
+        selectedAreas.isEmpty &&
+        selectedVersions.isEmpty) {
+      return List.from(entries);
```

Comment:
Does this need to return a new list? Doesn't look like callers should modify it.


```suggestion
      return entries.
```

If it does, then use a spread, `List.of`, or `.toList()` as `List.from` has some typing issues.


```suggestion
      return [...entries];
```